main/lsp/requests/type_definition.cc

```diff
@@ -0,0 +1,76 @@
+#include "common/typecase.h"
+#include "core/lsp/QueryResponse.h"
+#include "main/lsp/lsp.h"
+
+using namespace std;
+
+namespace sorbet::realmain::lsp {
+
+namespace {
+vector<core::Loc> locsForType(const core::GlobalState &gs, core::TypePtr type) {
+    vector<core::Loc> result;
+    typecase(
+        type.get(), [&](core::ClassType *t) { result.emplace_back(t->symbol.data(gs)->loc()); },
+        [&](core::AppliedType *t) { result.emplace_back(t->klass.data(gs)->loc()); },
+        [&](core::OrType *t) {
+            for (auto loc : locsForType(gs, t->left)) {
+                result.emplace_back(loc);
+            }
+            for (auto loc : locsForType(gs, t->right)) {
+                result.emplace_back(loc);
+            }
+        },
+        [&](core::AndType *t) {
+            for (auto loc : locsForType(gs, t->left)) {
+                result.emplace_back(loc);
+            }
+            for (auto loc : locsForType(gs, t->right)) {
+                result.emplace_back(loc);
+            }
+        });
+    return result;
+}
+} // namespace
+
+LSPResult LSPLoop::handleTextDocumentTypeDefinition(unique_ptr<core::GlobalState> gs, const MessageId &id,
+                                                    const TextDocumentPositionParams &params) const {
+    auto response = make_unique<ResponseMessage>("2.0", id, LSPMethod::TextDocumentTypeDefinition);
+    prodCategoryCounterInc("lsp.messages.processed", "textDocument.typeDefinition");
+    auto result = setupLSPQueryByLoc(move(gs), params.textDocument->uri, *params.position,
+                                     LSPMethod::TextDocumentTypeDefinition, false);
+    gs = move(result.gs);
+    if (result.error) {
+        // An error happened while setting up the query.
+        response->error = move(result.error);
+    } else {
+        auto &queryResponses = result.responses;
+        vector<unique_ptr<Location>> result;
+        if (!queryResponses.empty()) {
+            const bool fileIsTyped =
+                config.uri2FileRef(*gs, params.textDocument->uri).data(*gs).strictLevel >= core::StrictLevel::True;
+            auto resp = move(queryResponses[0]);
+
+            // Only support go-to-type-definition on constants in untyped files.
+            if (resp->isConstant() || (fileIsTyped && (resp->isIdent() || resp->isLiteral()))) {
+                auto tpo = resp->getTypeAndOrigins();
+                for (auto loc : locsForType(*gs, tpo.type)) {
+                    addLocIfExists(*gs, result, loc);
+                }
+            } else if (fileIsTyped && resp->isDefinition()) {
+                // TODO(jez) Not clear what go to type def on a definition should do...
```

Comment:
TypeScript just brings you to the definition site, so I think the go-to-def behavior works here too.